lexxpluss_apps/src/board_controller.cpp

```diff
@@ -0,0 +1,1384 @@
+                k_msgq_purge(&led_controller::msgq);
+        }
+    }
+    bool is_activated() const {
+        return activated;
+    }
+private:
+    int64_t start_time{0};
+    int thres{0}, counter{0};
+    bool activated{false};
+};
+
+#define PWS_LONG_PUSHED_MS 60000
+#define PWS_PUSHED_MS 3000
+class power_switch { // Variables Implemented
+public:
+    enum class STATE {
+        RELEASED, PUSHED, LONG_PUSHED,
+    };
+    void poll() {
+        gpio_dt_spec gpio_dev = GET_GPIO(ps_sw_in);
+        int now{gpio_pin_get_dt(&gpio_dev)};
+        if (prev_raw != now) {
+            prev_raw = now;
+            count = 0;
+        } else {
+            ++count;
+        }
+        bool asserted{false};
+        if (count > COUNT) {
+            count = COUNT;
+            asserted = true;
+        }
+        if (asserted) {
+            bool changed{prev != now};
+            sw_bat.poll(changed);
+            sw_unlock.poll(changed);
+            if (changed) {
+                prev = now;
+                start_time = k_uptime_get();
+            } else if (now == 0) {
+                auto elapsed{k_uptime_get() - start_time};
+                if (elapsed > PWS_LONG_PUSHED_MS) {
+                    if (state != STATE::LONG_PUSHED)
+                        state = STATE::LONG_PUSHED;
+                } else if (elapsed > PWS_PUSHED_MS) {
+                    if (state == STATE::RELEASED)
+                        state = STATE::PUSHED;
+                }
+            }
+        }
+    }
+    void reset_state() {
+        start_time = k_uptime_get();
+        state = STATE::RELEASED;
+    }
+    STATE get_state() const {return state;}
+    bool get_raw_state() {
+        gpio_dt_spec gpio_dev = GET_GPIO(ps_sw_in);
+        return gpio_pin_get_dt(&gpio_dev) == 0;
+    }
+    void set_led(bool enabled) {
+        gpio_dt_spec gpio_dev = GET_GPIO(ps_led_out);
+        gpio_pin_set_dt(&gpio_dev, enabled ? 0 : 1);
+    }
+    void toggle_led() {
+        set_led(led_en);
+        led_en = true ? false : true;
+    }
+    bool is_activated_battery() const {
+        return sw_bat.is_activated();
+    }
+    bool is_activated_unlock() const {
+        return sw_unlock.is_activated();
+    }
+private:
+    power_switch_handler sw_bat{2}, sw_unlock{10};
+    int64_t start_time;
+    STATE state{STATE::RELEASED};
+    uint32_t count{0};
+    bool led_en{false};
+    int prev{-1}, prev_raw{-1};
+    static constexpr uint32_t COUNT{1};
+};
+
+#define BUMPER_SWITCH_HOLD_TIME_MS 1000
+class bumper_switch { // Variables Implemented
+public:
+    void poll() {
+        if(asserted_flag) {
+            if ((k_uptime_get() - start_time) > BUMPER_SWITCH_HOLD_TIME_MS) {
+                asserted_flag = false;
+            }
+        } else {
+            gpio_dt_spec gpio_dev = GET_GPIO(bp_left);
+            if (gpio_pin_get_dt(&gpio_dev) == 0) {
+                asserted_flag = true;
+                start_time = k_uptime_get();
+            }
```

Comment:
I think it would be more correct to always run it, not within the else block.

The reason is that while the bumper is being pressed, we do not want it to return to normal, even if only momentarily.

elseブロック内ではなく、常に実行する方が正しいと思います。
バンパーが押されている間は、一瞬とはいえども正常に戻らないで欲しいと思うからです。

Reply:
過去の背景からすると、バンパーが押された瞬間にかなりチャタリングが発生したとのことでした。おそらくは一度assertを検出したら１秒間（おそらくチャタリングがおさまるまで）は状態を見に行かない、という意図からこのようなコードになっています。SCBとなって回路的にはフィルターが入っているのですが、同様の現象が発生する可能性があり、ここはこのままの方がよいのではないかと思っています。

I've heard from past background that there was quite a bit of chattering at the moment the bumper was pushed, so perhaps the code is this way with the intention of not going to look at the state for 1 second (perhaps until the chattering subsides) once the assert is detected. filter is in place, but I think it is better to leave it as it is here, as it may cause the same phenomenon.

Reply:
気にしているのは、バンパーが押されているのが1秒以上続いた時に、1周期だけ異常が解除される可能性です。

押されている間は1秒タイマをリセットし続けるようにすると、押されなくなってから1秒で異常解除になると考えています。

What I am concerned about is the possibility of the anomaly being lifted for one cycle when the bumper has been pressed for more than one second.

I'm thinking that if we keep resetting the 1-second timer as long as it is being pushed, the abnormality will be released 1 second after it is no longer being pushed.

Reply:
内容理解しました。そのようなケースがカバーできるようロジックを修正します。

Reply:
elseブロックを外しました。常時Bumperの状態をみて、assertされ続けていたらその間start_timeを更新し、releaseされてから1秒後に状態をreleaseにする、というロジックで問題なさそうです。

I removed the "else" block. The logic seems to work fine: always look at the status of the Bumper, update start_time if it keeps being asserted, and release the status 1 second after it is released.